pkg/volume/csi/csi_attacher.go

```diff
@@ -224,36 +243,55 @@ func (c *csiAttacher) Detach(volumeName string, nodeName types.NodeName) error {
 func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) error {
 	glog.V(4).Info(log("probing for updates from CSI driver for [attachment.ID=%v]", attachID))
 
-	ticker := time.NewTicker(c.waitSleepTime)
-	defer ticker.Stop()
+	watcher, err := c.k8s.StorageV1alpha1().VolumeAttachments().Watch(meta.SingleObject(meta.ObjectMeta{Name: attachID}))
+	if err != nil {
+		return fmt.Errorf("watch error:%v for volume %v", err, volumeHandle)
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
 
 	timeout := c.waitSleepTime * 10
 	timer := time.NewTimer(timeout) // TODO (vladimirvivien) investigate making this configurable
 	defer timer.Stop()
 
-	//TODO (vladimirvivien) instead of polling api-server, change to a api-server watch
+	attachmentExisted := false
+
 	for {
 		select {
-		case <-ticker.C:
-			glog.V(4).Info(log("probing VolumeAttachment [id=%v]", attachID))
-			attach, err := c.k8s.StorageV1alpha1().VolumeAttachments().Get(attachID, meta.GetOptions{})
-			if err != nil {
-				if apierrs.IsNotFound(err) {
-					//object deleted or never existed, done
-					glog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volumeHandle))
-					return nil
-				}
-				glog.Error(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
-				continue
+		case event, ok := <-ch:
+			if !ok {
+				glog.Errorf(log("[attachment.ID=%v] watch channel had been closed", attachID))
+				return errors.New("volume attachment watch channel had been closed")
 			}
 
-			// driver reports attach error
-			detachErr := attach.Status.DetachError
-			if detachErr != nil {
-				glog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
-				return errors.New(detachErr.Message)
+			switch event.Type {
+			case watch.Added, watch.Modified:
+				attachmentExisted = true
+				attach, _ := event.Object.(*storage.VolumeAttachment)
+				// driver reports attach error
+				detachErr := attach.Status.DetachError
+				if detachErr != nil {
+					glog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
+					return errors.New(detachErr.Message)
+				}
+			case watch.Deleted:
+				//object deleted
+				glog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] has been deleted", attachID, volumeHandle))
+				return nil
+
+			case watch.Error:
+				glog.Errorf(log("[attachment.ID=%v] watch error", attachID))
+				return errors.New("detacher.WaitForDetach failed for volume watch error")
 			}
+
 		case <-timer.C:
+			// if the volumeAttachment still exists, watch.Added will be triggered,and attachmentExisted will be true
+			if !attachmentExisted {
+				glog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volumeHandle))
```

Comment:
A simpler approach:
If the VolumeAttachment did existed, it should trigger an `watch.Add` event.
We do a GET here to confirm that the VolumeAttachment is not exist.
What do you think?
@saad-ali @caesarxuchao

Reply:
This sounds good to me.